crl/validator.go

```diff
@@ -20,346 +20,420 @@ package crl
-
-	for _, list := range v.lists {
-		listIssuerName := list.Issuer.String()
-
-		for _, cert := range list.RevokedCertificates {
-			if listIssuerName == issuer &&
-				cert.SerialNumber.Cmp(serialNumber) == 0 {
-				return true
-			}
+func (v *validator) syncLoop(ctx context.Context) {
+	ticker := time.NewTicker(syncInterval)
+	defer ticker.Stop()
+	v.sync() // first tick is after the interval
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			v.sync()
 		}
 	}
-
-	return false
 }
 
-func (v *validator) updateCRL(endpoint string) error {
-	v.listsLock.RLock()
-	crl, ok := v.lists[endpoint]
-	v.listsLock.RUnlock()
-
-	if !ok || crlHasExpired(crl, nowFunc()) {
-		return v.downloadCRL(endpoint)
+func (v *validator) validatorLoop(ctx context.Context) {
+	var f func()
+	var ok bool
+	for {
+		select {
+		case <-ctx.Done():
+			v.started = false
+			return
+		case f, ok = <-v.crlChan:
+			if ok {
+				f()
+			}
+		}
 	}
-
-	return nil
 }
 
-func (v *validator) verifyCRL(crl *x509.RevocationList) error {
-	v.certificatesLock.RLock()
-	defer v.certificatesLock.RUnlock()
-
-	issuerName := crl.Issuer.String()
-
-	certificate, ok := v.certificates[issuerName]
-	if !ok {
-		return errors.New("CRL signature could not be validated against known certificates")
+func (v *validator) Validate(certificate *x509.Certificate) error {
+	if !v.started {
+		return errors.New("CRL validator is not started")
 	}
-
-	return crl.CheckSignatureFrom(certificate)
+	return v.validateCert(certificate)
 }
 
-func (v *validator) downloadCRL(endpoint string) error {
-	response, err := v.httpClient.Get(endpoint)
-	if err != nil {
-		return err
+func (v *validator) SetValidatePeerCertificateFunc(config *tls.Config) error {
+	if v.started {
+		return errors.New("SetVerifyPeerCertificateFunc must be called before Start")
 	}
 
-	defer response.Body.Close()
-
-	data, err := io.ReadAll(response.Body)
-	if err != nil {
-		return fmt.Errorf("unable to download CRL (url=%s): %w", endpoint, err)
-	}
-
-	crl, err := x509.ParseRevocationList(data)
-	if err != nil {
-		return fmt.Errorf("unable to parse downloaded CRL (url=%s): %w", endpoint, err)
+	parseCertificatesSlice := func(rawCerts [][]byte) ([]*x509.Certificate, error) {
+		var raw []byte
+		for _, rawCert := range rawCerts {
+			raw = append(raw, rawCert...)
+		}
+		return x509.ParseCertificates(raw)
 	}
 
-	if err := v.verifyCRL(crl); err != nil {
-		return fmt.Errorf("CRL verification failed (issuer=%s): %w", crl.Issuer.String(), err)
+	// check that all cert issuers are in the truststore, and add missing crl distribution points.
+	for _, chain := range config.Certificates {
+		certificates, err := parseCertificatesSlice(chain.Certificate)
+		if err != nil {
+			return err
+		}
+		for _, cert := range certificates {
+			issuer := v.truststore[cert.Issuer.String()]
+			if issuer == nil {
+				// This indicates a mismatch between crlValidator truststore and tls.Config. This is a programming error.
```

Comment:
It could be the truststore is missing a certificate, which is a configuration error?

In that case, we might have to revisit the error message. Or, the truststore should check it. But if it starts failing because an unrelated cert (maybe without CRL) is missing, it might be a breaking change for some parties?

Reply:
We use the same truststore for the crl validator and network layer. So this can never fail until we deviate from this pattern. At this point the CRLs behavior should be reevaluated. So this is prevents a future bug.

Reply:
Ahh, I see your point. We probably need a validation step in the trust configuration to see if the provided cert form a chain to some root?

Reply:
https://github.com/nuts-foundation/nuts-node/issues/2025